### PR TITLE
♻️🚀 Dead-code-eliminate `ShadowCSS.js`

### PR DESCRIPTION
**Pull request by alanorozco**

### CODE REVIEW

File: third_party/webcomponentsjs/ShadowCSS.js [hunk 1]
```diff
@@ -121,152 +121,178 @@
...
-          // https://connect.microsoft.com/IE/feedbackdetail/view/955703/accessing-csstext-of-a-keyframe-rule-that-contains-a-webkit-property-via-cssom-generates-exception
-          try {
-            if (rule.cssText) {
-              cssText += rule.cssText + '\n\n';
-            }
-          } catch(x) {
-            if (rule.type === CSSRule.KEYFRAMES_RULE && rule.cssRules) {
-              cssText += this.ieSafeCssTextFromKeyFrameRule(rule);
-            }
+// [@ampproject difference with original]
```
Comment (alanorozco) [inline third_party/webcomponentsjs/ShadowCSS.js:124]: I recommend disabling whitespace to read this diff. [Click here to disable](&w=1) or use the dropdown on ⚙️:

image

Comment (alanorozco) [inline third_party/webcomponentsjs/ShadowCSS.js:124]: I recommend disabling whitespace to read this diff.

[Click here to disable](&w=1) or use the dropdown on ⚙️:

image

**Comment by alanorozco** [comment]
@samouri It's also `amp4ads-v0.js` and [all 20 of `amp-story`-related binaries](https://github.com/ampproject/amphtml/pull/38013/checks?check_run_id=5828026779), some of them being smaller by >6.5%.

Note that the last we time changed this file was [nearly six years ago](https://github.com/ampproject/amphtml/commit/d084ee388ea8dd5db847ba37b3aa161716bfd15a), and we've considered it a fork of the original since then.

File: third_party/webcomponentsjs/ShadowCSS.js (hunk 1 above)
Comment (jridgewell) [reply third_party/webcomponentsjs/ShadowCSS.js:124]: This belongs in the `README.amp` file in this directory

File: third_party/webcomponentsjs/ShadowCSS.js [hunk 2]
```diff
@@ -121,152 +121,158 @@
...
+  }
+  var re = makeScopeMatcher(scopeSelector);
+  return !selector.match(re);
+}
+/** @this {ShadowCSS} */
+export function makeScopeMatcher(scopeSelector) {
+  scopeSelector = scopeSelector.replace(/\[/g, '\\[').replace(/\]/g, '\\]');
+  return new RegExp('^(' + scopeSelector + ')' + selectorReSuffix, 'm');
+}
+/** @this {ShadowCSS} */
```
Comment (samouri) [inline third_party/webcomponentsjs/ShadowCSS.js:206]: FMI: what is the significance of the `@this` documentation?

File: third_party/webcomponentsjs/README.amp [hunk 3]
```diff
@@ -8,3 +8,14 @@ Shadow DOM-related polyfills
 Local Modifications:
 - Stripped to only perform necessary Shadow CSS polyfilling.
 - Shortened License based on intended functionality from build pipeline.
+- Added feature switches that allows certain codepaths to be
+  dead-code-eliminated: `allowArraySelectors` and `isIeSupported`
```
Comment (samouri) [inline third_party/webcomponentsjs/README.amp:12]: FMI: how do you know we don't use array selectors?

File: third_party/webcomponentsjs/ShadowCSS.js [hunk 4]
```diff
@@ -121,152 +121,158 @@
...
+        // if this happens, we fallback to constructing the rule
+        // from the CSSRuleSet
+        // https://connect.microsoft.com/IE/feedbackdetail/view/955703/accessing-csstext-of-a-keyframe-rule-that-contains-a-webkit-property-via-cssom-generates-exception
+        try {
+          if (rule.cssText) {
+            cssText += rule.cssText + '\n\n';
+          }
+        } catch(x) {
+          if (
+            isIeSupported &&
```
[the remark this answers is not in the document]
Comment (alanorozco) [reply third_party/webcomponentsjs/ShadowCSS.js:155]: This is a `CSSRuleList`, which lacks `forEach`. Other changes in syntax are minimal, not really worth doing.

**Review by samouri — APPROVED** [review]

**Review by jridgewell — APPROVED** [review]

**Pull request merged**